parser.rst

```diff
@@ -0,0 +1,903 @@
+parsed.
+
+Left recursion
+~~~~~~~~~~~~~~
+
+PEG parsers normally do not support left recursion but CPython's parser
+generator implements a technique similar to the one described in Medeiros et al.
+[2]_ but using the memoization cache instead of static variables. This approach
+is closer to the one described in Warth et al. [3]_. This allows us to write not
+only simple left-recursive rules but also more complicated rules that involve
+indirect left-recursion like::
+
+  rule1: rule2 | 'a'
+  rule2: rule3 | 'b'
+  rule3: rule1 | 'c'
+
+and "hidden left-recursion" like::
+
+  rule: 'optional'? rule '@' some_other_rule
+
+Variables in the Grammar
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+A sub-expression can be named by preceding it with an identifier and an
+``=`` sign. The name can then be used in the action (see below), like this: ::
+
+    rule_name[return_type]: '(' a=some_other_rule ')' { a }
+
+Grammar actions
+~~~~~~~~~~~~~~~
+
+.. _peg-grammar-actions:
+
+To avoid the intermediate steps that obscure the relationship between the
+grammar and the AST generation the PEG parser allows directly generating AST
+nodes for a rule via grammar actions. Grammar actions are language-specific
+expressions that are evaluated when a grammar rule is successfully parsed. These
+expressions can be written in Python or C depending on the desired output of the
+parser generator. This means that if one would want to generate a parser in
+Python and another in C, two grammar files should be written, each one with a
+different set of actions, keeping everything else apart from said actions
+identical in both files. As an example of a grammar with Python actions, the
+piece of the parser generator that parses grammar files is bootstrapped from a
+meta-grammar file with Python actions that generate the grammar tree as a result
+of the parsing. 
+
+In the specific case of the PEG grammar for Python, having actions allows
+directly describing how the AST is composed in the grammar itself, making it
+more clear and maintainable. This AST generation process is supported by the use
+of some helper functions that factor out common AST object manipulations and
+some other required operations that are not directly related to the grammar.
+
+To indicate these actions each alternative can be followed by the action code
+inside curly-braces, which specifies the return value of the alternative::
+
+    rule_name[return_type]:
+        | first_alt1 first_alt2 { first_alt1 }
+        | second_alt1 second_alt2 { second_alt1 }
+
+If the action is omitted and C code is being generated, then there are two
+different possibilities:
+
+1. If there’s a single name in the alternative, this gets returned.
+2. If not, a dummy name object gets returned (this case should be avoided).
+
+If the action is omitted and Python code is being generated, then a list
+with all the parsed expressions gets returned (this is meant for debugging).
```

Comment:
I don't think that's true anymore. I think I remember Matthieu (I guess the state of the python generator in the pegen repo is going to be the standard for this document as well as far as the Python generator goes) changing this to most closely resemble what the C generator does: If there's a single name it gets returned, otherwise the list with all parsed expressions gets returned. Do you think we should change it?

Reply:
Yeah, let's change it to the current behaviour

Reply:
Also, I still want to keep the Python generator in the CPython repo updated with the one in pegen